### PR TITLE
Add fsgroup mount proposal

### PR DESCRIPTION
**Pull request by gnufied**
xref https://github.com/kubernetes/enhancements/issues/2317

/sig storage

### CODE REVIEW

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+## Proposal
+
+We are updating CSI specs by adding additional field called `volume_mount_group` to `NodeStage` and `NodePublish` RPC calls. The CSI proposal is available at - https://github.com/container-storage-interface/spec/pull/468 .
+
+The CSI spec change is deliberately trying to avoid asking drivers to use supplied `fsGroup` as a generic handle for ownership and permissions. The reason being - Kubernetes may expect ownership and permissions to be in a way that is very platform/OS specific. We do not think CSI driver is right place to enforce all kind of different permissions expected by Kubernetes. The full scope of that discussion is out of scope for this enhancement and interested folks can follow along on - https://github.com/container-storage-interface/spec/issues/449
+
+
+### Risks and Mitigations
+
+I am not aware of any associated risks. If a driver can not support using `fsgroup` as a mount option, it can always use `FileFSGroupPolicy` and let kubelet handle the ownership and permissions.
```
Comment (jsafrane) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:72]: Please elaborate somewhere that there is a behavioral change: if two pods use the same volume with different `fsGroup`, then:
* Before this change, the pod that starts as second will `chown` all files, so the first pod probably can't access them.
* After this change, both pods may access the volume, if the volume can be mounted twice with different options.

This change is neligible and most probably corresponds to what in-tree Azure File plugin already does today.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+
+-->
+
+### Feature Enablement and Rollback
+
+_This section must be completed when targeting alpha to a release._
```
Comment (jsafrane) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:239]: Please file this section, it's required for alpha

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Currently for most volume plugins kubelet applies fsgroup ownership and permission based changes by recursively `chown`ing and `chmod`ing the files and directories inside a volume. For certain CSI drivers this may not be possible because `chown` and `chmod` are unix primitives and underlying CSI driver
+may not support them. This enhancement proposes providing the CSI driver with fsgroup as an explicit field so as CSI driver can apply this on mount time.
+
+## Motivation
+
+Since some CSI drivers(AzureFile for example) - don't support chmod/chown - we propose that fsgroup of pod to be provided to the CSI driver on `NodeStage`
```
Comment (jsafrane) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:51]: I would emphasize that this feature is a prerequisite for CSI migration of AzureFile volume plugin and removal of Azure cloud provider

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+When Kubelet sees `CSIDriver.Spec.FSGroupPolicy` value set to `Mount` - it would use proposed CSI field `volume_mount_group` to pass pod's `fsGroup` to the CSI driver. Kubelet will expect that driver will use
+this field for mounting volume with given `fsGroup` and no further permission/ownerhip change will be necessary.
+
+### Test Plan
+
+Unit test:
+1. Test that whenever supported pod's `fsGroup` should be passed to CSI driver via `volume_mount_group` field.
+
+For alpha feature:
+1. Run manual tests against azurefile CSI driver.
```
Comment (jsafrane) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:97]: The driver needs to be updated too.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+### Graduation Criteria
+
+<!--
+**Note:** *Not required until targeted at a release.*
```
Comment (jsafrane) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:124]: Please file this section

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 1 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:72]: No before this change - no chown/chmod will be performed at all and hence except the root user(or pods that run as root),  no pod will be able to read/write to the volume. 

I am not sure if this is a behaviour change. Before - only root could read/write to the pod and now other users can too.  I could add something here but I am not sure if it fits under risks and mitigations.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+    ReadWriteOnceWithFSTypeFSGroupPolicy FSGroupPolicy = "ReadWriteOnceWithFSType"
+    FileFSGroupPolicy FSGroupPolicy = "File"
+    NoneFSGroupPolicy FSGroupPolicy = "None"
+    // Indicates that the CSI driver supports applying fsgroup via mount option
+    MountFSGroupPolicy FSGroupPolicy = "Mount"
+)
+```
+
+When Kubelet sees `CSIDriver.Spec.FSGroupPolicy` value set to `Mount` - it would use proposed CSI field `volume_mount_group` to pass pod's `fsGroup` to the CSI driver. Kubelet will expect that driver will use
+this field for mounting volume with given `fsGroup` and no further permission/ownerhip change will be necessary.
```
[the remark this answers is not in the document]
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md]: I removed this section. I thought long and hard about keeping `FSGroupPolicy` field because initially we weren't sure how CSI spec change will look like.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 2 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md]: updated.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 3 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:53]: fixed

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 4 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:90]: fixed

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 5 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md]: kinda fixed. PTAL

**Comment by gnufied** [comment]
cc @andyzhangx

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+  Enabling this feature on its own does not changes any existing behaviour but if
+  selected CSI driver advertise `VOLUME_MOUNT_GROUP` node capability, it will result in
+  volume being mounted with pod's `fsGroup`.
```
Comment (deads2k) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:205]: who sets this value?  End-user, correct?

How does an end user know if his setting took effect?

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  Yes - feature gate can be disabled once enabled and kubelet will fallback to its current behavior, which will
+  result in any CSI driver that requires using `fsGroup` as a mount option to not use the mount option. In which case
+  mounted volumes will be unwritable by any pods other than those running as root.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+  It will cause kubelet to use `volume_mount_group` field of CSI whenver applicable as discussed in above design.
```
Comment (deads2k) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:214]: would pods would have to be recreated or woudl a crashlooping pod suddenly start working again?

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+
+* **Are there any tests for feature enablement/disablement?**
+  Not yet.
+
+### Rollout, Upgrade and Rollback Planning
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can a rollout fail? Can it impact already running workloads?**
```
Comment (deads2k) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:223]: This question seems important for this feature. Does the scheduler have any way to know whether a particular node will honor this part of the podspec?

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+  Try to be as paranoid as possible - e.g., what if some components will restart
+   mid-rollout?
+
+* **What specific metrics should inform a rollback?**
```
Comment (deads2k) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:227]: I think you'll want to list the metrics that you use for determining successful and failed mounts here.  And probably count cases where the pod's intent wasn't fulfilled.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+
+* **Is the rollout accompanied by any deprecations and/or removals of features, APIs,
+fields of API types, flags, etc.?**
+  Even if applying deprecation policies, they may still surprise some users.
+
+### Monitoring Requirements
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can an operator determine if the feature is in use by workloads?**
```
Comment (deads2k) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:242]: see my metric comment above.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+* **Are there any missing metrics that would be useful to have to improve observability
+of this feature?**
+  Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
+  implementation difficulties, etc.).
+
+### Dependencies
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **Does this feature depend on any specific services running in the cluster?**
```
Comment (deads2k) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:274]: If I have understood correctly, this relies on
1. the user to set the value in the pod
2. the kubelet to call the options with the featuregate set
3. the CSI driver to support the  option.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 7 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:205]: yes `fsGroup` is being set by the enduser in their pods. The only way for a enduser to know if setting took effect is by execing into pod and checking permissions of mounted volume (or have some healthcheck script which does same).

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 8 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:214]: Yes pods have to be re-created for the setting to take effect.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 9 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:223]: I do not think scheduler has to know and it can be pretty expensive(code and complexity perspective) to bubble this up to scheduler.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Currently for most volume plugins kubelet applies fsgroup ownership and permission based changes by recursively `chown`ing and `chmod`ing the files and directories inside a volume. For certain CSI drivers this may not be possible because `chown` and `chmod` are unix primitives and underlying CSI driver
+may not support them. This enhancement proposes providing the CSI driver with fsgroup as an explicit field so as CSI driver can apply this on mount time.
+
+## Motivation
+
+Since some CSI drivers(AzureFile for example) - don't support chmod/chown - we propose that fsgroup of pod to be provided to the CSI driver on `NodeStage`
```
Comment (xing-yang) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:53]: NodeStage -> NodeStageVolume

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+and `NodePublish` CSI RPC calls. This will allow the CSI driver to apply `fsgroup` as a mount option on `NodeStage` or `NodePublish` and kubelet can be freed
```
Comment (xing-yang) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:54]: NodePublish -> NodePublishVolume

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+
+- Allow CSI driver to mount volumes with provided fsgroup.
+
+### Non-Goals
+
+- We are not supplying `fsGroup` as a generic ownerhip and permission handle to the CSI driver. We do not expect CSI drivers to `chown` or `chmod` files.
+
+## Proposal
+
+We are updating CSI specs by adding additional field called `volume_mount_group` to `NodeStage` and `NodePublish` RPC calls. The CSI proposal is available at - https://github.com/container-storage-interface/spec/pull/468 .
```
Comment (xing-yang) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:69]: `NodeStage` and `NodePublish` -> `NodeStageVolume` and `NodePublishVolume`

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+### Risks and Mitigations
+
+I am not aware of any associated risks. If a driver can not support using `fsgroup` as a mount option, it can always use `FileFSGroupPolicy` and let kubelet handle the ownership and permissions.
+
+## Design Details
+
+We are proposing that when kubelet determines a CSI driver has `VOLUME_MOUNT_GROUP` node capability, the kubelet will use proposed CSI field `volume_mount_group` to pass pod's `fsGroup` to the CSI driver. Kubelet will expect that driver will use
+this field for mounting volume with given `fsGroup` and no further permission/ownerhip change will be necessary.
+
+It should be noted that if a CSI driver advertises `VOLUME_MOUNT_GROUP` node capability then value defined in `CSIDriver.Spec.FSGroupPolicy` will be ignored and kubelet willl always use `fsGroup` as a mount option.
```
Comment (xing-yang) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:83]: typo: willl -> will

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+
```
Comment (xing-yang) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:84]: If a CSI driver does not make any changes proposed in this KEP, will there be any behavior change?  Please clarify that in the KEP.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
+### Test Plan
+
+Unit test:
+1. Test that whenever supported pod's `fsGroup` should be passed to CSI driver via `volume_mount_group` field.
+
+For alpha feature:
+1. Update Azure File CSI driver to support supplying `fsGroup` via `NodeStage` and `NodePublish`.
```
Comment (xing-yang) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:91]: There should be tests for non-Azure File CSI driver to make sure there are no regressions.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 19]
```diff
@@ -0,0 +1,400 @@
...
+1. Run manual tests against azurefile CSI driver.
+
+For beta:
+1. E2E tests that verify volume readability/writability using azurefile CSI driver.
+2. E2E tests using CSI mock driver.
```
Comment (xing-yang) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:96]: Should there be E2E tests for CSI hostpath driver and GCE PD driver as well?

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 20]
```diff
@@ -0,0 +1,400 @@
...
+
+#### Alpha -> Beta Graduation
+
+- Since this feature is a must-have for azurefile CSI migration, we will perform testing of the driver.
+- Currently CSI spec change is being introduced as alpha change and we will work to move the API change in CSI spec to stable.
+
+#### Beta -> GA Graduation
+
+- CSI spec change should be stable.
+- Tested via e2e and manually using azurefile CSI driver.
```
Comment (xing-yang) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:126]: Testing no regressions are introduced in non-azurefile CSI driver?

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 21]
```diff
@@ -0,0 +1,400 @@
...
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- Since this feature is a must-have for azurefile CSI migration, we will perform testing of the driver.
```
Comment (xing-yang) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:120]: Testing no regressions are introduced in non-azurefile CSI driver?

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 17 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:84]: it is somewhat covered in section - "* **Does enabling the feature change any default behavior?**"

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 18 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:91]: There are already tests that verify fsgroup behaviour for existing drivers.  I do not think we need specifically new tests for them - I think.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 19 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:96]: I do not think so. As I mentioned above - there are already fsgroup based tests for those drivers.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 20 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:126]: same as above. we already have fsgroup tests that verify behaviour of drivers that don't have this new capability.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 21 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:120]: same as above

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 15 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md]: fixed

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 10 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:227]: Do we need these sections for alpha feature? The PRR check says:

> _This section must be completed when targeting beta graduation to a release._

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 11 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:242]: same as above.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 10 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:225]: Updated this section. However kubelet never tracked whether pod's intent was fulfilled and I think such a metric is out of scope for this specific change because such a metric will apply to generic mechanism by which `fsGroup` is applied to the volumes (and they are in-use since k8s <1.2 or something)

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 11 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:242]: Updated.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 12 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:272]: added

**Comment by gnufied** [comment]
@deads2k @xing-yang I have addressed  the review comments. PTAL. thanks.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 7 above)
Comment (deads2k) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:205]: > yes `fsGroup` is being set by the enduser in their pods. The only way for a end-user to know if setting took effect is by execing into pod and checking permissions of mounted volume (or have some healthcheck script which does same).

This seems like the sort of thing that users would have difficulty understanding the failure mode for.  Do you have recommended scripts or something to recommend to a user that wants to be sure his files are correct?

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 8 above)
Comment (deads2k) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:214]: > Yes pods have to be re-created for the setting to take effect.

heh, don't tell me, write it in the doc :)

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 7 above)
Comment (deads2k) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:205]: Also, can we get this updated in the doc?  I won't be the only admin and/or user wondering.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 9 above)
Comment (deads2k) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:221]: > I do not think scheduler has to know and it can be pretty expensive(code and complexity perspective) to bubble this up to scheduler.

So that means that mid-rollout (either turning it on or turning it off) can result in different pods of the same replicaset/deployment failing or succeeding in seemingly random ways?  I think that's worth noting in an answer to this question.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 10 above)
Comment (deads2k) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:225]: > Updated this section. However kubelet never tracked whether pod's intent was fulfilled and I think such a metric is out of scope for this specific change because such a metric will apply to generic mechanism by which `fsGroup` is applied to the volumes (and they are in-use since k8s <1.2 or something)

I can see an argument for not requiring it for alpha, but I don't think I follow the reasoning that because there was no metric before there shouldn't be a metric as this feature moves to beta.  Without a metric, how can you know that
1. the feature is in use
2. the feature is working.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 12 above)
Comment (deads2k) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:272]: > added

I'm not seeing it in this sectin.

Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:272]: It is there:

```
This feature depends on presence of `fsGroup` field in Pod and CSI driver having `VOLUME_MOUNT_GROUP`
  capability.

  - [CSI driver]
    - CSI driver must have `VOLUME_MOUNT_GROUP` capability:
      - If this capability is not available in CSI driver then kubelet will try to use default mechanism of applying `fsGroup` to volume (which is basically `chown` and `chomid`). If underlying driver however does not support applying group permissions via `chown` and `chmod` then Pods will not run correctly.
      - Workloads may not run correctly or volume has to be mounted with permissions `777`.

```

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 20 above)
Comment (xing-yang) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:126]: Can you add a link to the existing e2e tests https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/fsgroupchangepolicy.go to this KEP here?

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 9 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:221]: hmm - if the replicaset was say working in 1.20 - then user must be running the pods as root or mounting the volume as world-writable (`777`). So this feature won't affect that. Even if `fsGroup` is applied correct to the volume, it won't affect anything that is running as root (because those pods can write to any volume anyways) or volumes with 777 permissions (because those volumes can be written by anyone anyways).

So in a nutshell - it does not have any impact of already correctly running workloads. Now if you somehow had a pod in 1.20 - which was not running as root and volumes weren't mounted with 777 permissions then those would suddenly start working. But I do not think that is the use case this section documents.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 22]
```diff
@@ -0,0 +1,393 @@
...
+
+* **How can an operator determine if the feature is in use by workloads?**
+  Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+  checking if there are objects with field X set) may be a last resort. Avoid
+  logs or events for this purpose.
+
+* **What are the SLIs (Service Level Indicators) an operator can use to determine
+the health of the service?**
+  - [ ] Metrics
+    - Metric name:
```
Comment (xing-yang) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:250]: Can you fill out this section?

Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:250]: This section is only needed for beta features as far as I can tell. We are going to do some work to split some of the metrics - https://github.com/kubernetes/kubernetes/issues/98667 . The SLIs could be determined in next release I think (when we know more about real world usage).

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 8 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md]: done

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 7 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:191]: I have updated the doc about what admin/pod owner should expect here.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 10 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:218]: sure.  Added more information about metrics and filed https://github.com/kubernetes/kubernetes/issues/98667 to capture how `fsGroup` intent was full-filled.

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 20 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:112]: done.

**Comment by deads2k** [comment]
/approve

PRR

File: keps/sig-storage/2317-fsgroup-on-mount/README.md [hunk 23]
```diff
@@ -0,0 +1,384 @@
...
+      - Kubelet
+  - [ ] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+* **Does enabling the feature change any default behavior?**
+  Enabling this feature on its own does not changes any existing behaviour but if
```
Comment (xing-yang) [inline keps/sig-storage/2317-fsgroup-on-mount/README.md:189]: nit: does not changes -> does not change

**Comment by xing-yang** [comment]
LGTM

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 1 above)
Comment (jsafrane) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:76]: A CSI driver could have supported `chown` / `chmod` for RWO / RWX volumes (such as github.com/kubernetes-csi/csi-driver-smb, if the samba export is configured properly), it's just faster with a mount option. Users of this driver could experience a very small behavior change.

Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md:76]: addressed under the section - `Does enabling the feature change any default behavior?**`

File: keps/sig-storage/2317-fsgroup-on-mount/README.md (hunk 23 above)
Comment (gnufied) [reply keps/sig-storage/2317-fsgroup-on-mount/README.md]: this line is replaced entirely now.
